### PR TITLE
16 cryptography models libraries and queries migration

### PR DESCRIPTION
**Pull request by ropwareJB**
Rebased https://github.com/microsoft/codeql/pull/17 onto upstream/main.

@bdrodes:
> Migration of C/C++ crypto models, inventory queries and example alerts.

### CODE REVIEW

**Review by RasmusWL — APPROVED** [review]
File layout :+1: for Python (did not review actual QL code yet)

Thanks for removing security tags for now, we should remember to add them when we want to publish these queries :sweat_smile: (I requested this to happen offline)

**Review by jketema — COMMENTED** [review]
The C++ file-layout generally looks ok to me. However, the split between `inventory/new_models` and `inventory/old_models` eludes me, especially since `inventory/old_models` is empty. What is the purpose of this?

Also per our [contributing guide](https://github.com/github/codeql/blob/main/CONTRIBUTING.md), all the queries will need to be tagged with the `experimental` tag, which seems to be missing ("Experimental queries need to include `experimental` in their `@tags`").

**Comment by bdrodes** [comment]
> The C++ file-layout generally looks ok to me. However, the split between `inventory/new_models` and `inventory/old_models` eludes me, especially since `inventory/old_models` is empty. What is the purpose of this?
> 
> Also per our [contributing guide](https://github.com/github/codeql/blob/main/CONTRIBUTING.md), all the queries will need to be tagged with the `experimental` tag, which seems to be missing ("Experimental queries need to include `experimental` in their `@tags`").
> 
> I'm also slightly concerned by the lack of tests.

The new vs old models exists to keep consistency with each language. For CPP we don't have old models, but I also haven't done a deep dive to try to artificially create some. It's also possible I missed some. 

The tags were removed by request to avoid automatic pack inclusion. 

We have tests for some of sample alerts internally to Microsoft, but the point of these was to be examples, but we can definitely get tests for these. In terms of the CBOM queries, for this initial push, the focus was on getting our queries into a public space. I expect the underlying models will change and what is being alerted on might change in the near future as well, so our intent was that tests would be developed in the near future for these queries. In fact, we are currently figuring out what test means for bill of material queries as part of this effort (@pwntester is currently running cbom queries on various public repos). My thought was a test should compare the new vs old models so we have something to compare against for regression, which I believe is what @pwntester is currently experimenting with.

**Comment by jketema** [comment]
> The new vs old models exists to keep consistency with each language. For CPP we don't have old models, but I also haven't done a deep dive to try to artificially create some. It's also possible I missed some.

That doesn't really tell me a lot. What's the purpose the the "old" models? Taking the word "old" face-value it's sound like something out dated that we do not want in the repo.

**Comment by bdrodes** [comment]
> > The new vs old models exists to keep consistency with each language. For CPP we don't have old models, but I also haven't done a deep dive to try to artificially create some. It's also possible I missed some.
> 
> That doesn't really tell me a lot. What's the purpose the the "old" models? Taking the word "old" face-value it's sound like something out dated that we do not want in the repo.

In some languages there are existing models for cryptography, either partial or mature. Where those exist, we develop queries for CBOM using those models we call them "old" models compared to the current PQC bill of material effort. Queries built upon any new modeling for that purposes of PQC we are calling "new".

**Comment by jketema** [comment]
> > > The new vs old models exists to keep consistency with each language. For CPP we don't have old models, but I also haven't done a deep dive to try to artificially create some. It's also possible I missed some.
> > 
> > 
> > That doesn't really tell me a lot. What's the purpose the "old" models? Taking the word "old" face-value it's sound like something out dated that we do not want in the repo.
> 
> In some languages there are existing models for cryptography, either partial or mature. Where those exist, we develop queries for CBOM using those models we call them "old" models compared to the current PQC bill of material effort. Queries built upon any new modeling for that purposes of PQC we are calling "new".

That still doesn't help a lot. Are the old models ours or yours? If they're ours they will be located somewhere else already, and a copy seems unnecessary. If they're yours, why should we become responsible for maintaining something that still sounds like it's outdated.

**Comment by jketema** [comment]
Thanks for the clarification. I think there's a discussion to be had - at a later moment - as to whether we should have all these inventory queries, but the file-layout looks good as-is.

File: cpp/ql/lib/experimental/cryptography/Concepts.qll [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import experimental.cryptography.CryptoArtifact
+import experimental.cryptography.CryptoAlgorithmNames
+
+import experimental.cryptography.modules.OpenSSL as OpenSSL
```
Comment (aeisenberg) [inline cpp/ql/lib/experimental/cryptography/Concepts.qll:3]: For consistency with other names, can you change this to `OpenSsl

File: cpp/ql/lib/experimental/cryptography/CryptoArtifact.qll [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+import semmle.code.cpp.ir.dataflow.TaintTracking
+
+/*
+ * A cryptographic artifact is a DataFlow::Node associated with some
+ * operation, algorithm, or any other aspect of cryptography.
+ */
+
+abstract class CryptographicArtifact extends Expr { }
+
+// /**
```
Comment (aeisenberg) [inline cpp/ql/lib/experimental/cryptography/CryptoArtifact.qll:12]: Any reason why this is commented out? Can you please include a comment explaining why, and why it is important to leave in, or else delete it.

File: cpp/ql/lib/experimental/cryptography/CryptoArtifact.qll [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+ *
+ * In principle, this class should model the location where an algorithm enters the program, not
+ * necessarily where it is used.
+ */
+abstract class CryptographicAlgorithm extends CryptographicArtifact {
+  abstract string getName();
+
+  abstract string getAlgType();
+
+  //  string getAlgType(){
```
Comment (aeisenberg) [inline cpp/ql/lib/experimental/cryptography/CryptoArtifact.qll:113]: Same as above.

File: cpp/ql/lib/experimental/cryptography/modules/OpenSSL.qll [hunk 4]
```diff
@@ -0,0 +1,718 @@
+import cpp
+import experimental.cryptography.CryptoAlgorithmNames
+import experimental.cryptography.CryptoArtifact
+import experimental.cryptography.utils.OpenSSL.CryptoFunction
+import experimental.cryptography.utils.OpenSSL.AlgorithmSink
+import experimental.cryptography.utils.OpenSSL.PassthroughFunction
+import experimental.cryptography.utils.OpenSSL.CryptoAlgorithm
+import experimental.cryptography.CryptoArtifact
+// import semmle.code.cpp.ir.dataflow.TaintTracking
```
Comment (aeisenberg) [inline cpp/ql/lib/experimental/cryptography/modules/OpenSSL.qll:9]: Delete

**Review by aeisenberg — COMMENTED** [review]
I'm generally fine with this. I haven't looked at the queries in detail. I'll trust the review of @RasmusWL and @jketema.

There is a lot of commented out code. I'd prefer that either the code is deleted, or there is a clear reason why the commented code should remain.

Also, there is a code scanning warning about using camel case of OpenSSL. I'm find either way.

**Review by aeisenberg — APPROVED** [review]
Approving, assuming that the code can be cleaned up in a followup.

**Review by RasmusWL — APPROVED** [review]
As I said [previously](https://github.com/github/codeql/pull/14289#pullrequestreview-1648094312)

> File layout 👍 for Python (did not review actual QL code yet)
>
> Thanks for removing security tags for now, we should remember to add them when we want to publish these queries 😅 (I requested this to happen offline)

**Pull request merged**